### PR TITLE
Rule S4226: Extensions should be in separate namespaces

### LINKED ISSUE #741: Rule S4226: Extensions should be in separate namespaces
[RSPEC-4226](https://jira.sonarsource.com/browse/RSPEC-4226)

It makes little sense to create an extension method when it is possible to just add that method to the type itself.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #741

### CODE REVIEW

**Comment by Evangelink** [comment]
Quite a lot of issues are raised but they all comply to the rule (i.e. a separate namespace for the extension method and the type being extended).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExtensionMethodShouldBeInSeparateNamespace.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+                    var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+
+                    if (methodSymbol != null &&
+                        methodSymbol.IsExtensionMethod &&
+                        methodSymbol.Parameters.Length > 0 &&
+                        methodSymbol.ContainingNamespace == methodSymbol.Parameters[0].Type.ContainingNamespace)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExtensionMethodShouldBeInSeparateNamespace.cs]: Could Type be empty? What if the type is ErrorType (or somethine else weird)?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+using System;
+
+namespace SomeNamespace
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs]: Let's add tests for global namespace. Mainly for documentation purposes, as I think the code should work fine.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExtensionMethodShouldBeInSeparateNamespace.cs (hunk 1 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExtensionMethodShouldBeInSeparateNamespace.cs]: Good catch.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs (hunk 2 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs]: Good catch.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+{
+    class Program
+    {
+    }
+
+    static class Helpers1
+    {
+        static void Foo(this Program p) // Noncompliant {{Either move this extension to another namespace or move the method inside the type itself.}}
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs]: The message will not be always correct - if you are extending an interface, you cannot move the method inside (yet). I would rephrase to something like "Move this extension method in another namespace." when extending interface or type that has no declaration (e.g. 3rd party class for example), and "Either move this extension method to another namespace or declare it as an instance method in 'Program'." when extending class.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs (hunk 2 above)
Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldBeInSeparateNamespace.cs]: Another corner case: extension methods in nested classes.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/Rules.Description/S4226.html [hunk 4]
```diff
@@ -0,0 +1,39 @@
+<p>It makes little sense to create an extension method when it is possible to just add that method to the type itself.</p>
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/Rules.Description/S4226.html]: This justification doesn't apply if (1) the type is defined in a different assembly, or (2) the type is an interface.
I'm not saying that the rule shouldn't fire in these cases, just that the current wording will make it look like a false positive to the user.

Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Utilities/Rules.Description/S4226.html]: Extension methods can only be declared in top-level classes (had do double-check, compiler reports error CS1109).

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExtensionMethodShouldBeInSeparateNamespace.cs (hunk 1 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExtensionMethodShouldBeInSeparateNamespace.cs]: If the `Type` is `ErrorType` the namespace is global namespace. But I will keep the test.
